port skat

patrick-schultz opened the pull request.

jigold commented on python/hail/expr/functions.py:
```diff
@@ -382,6 +382,24 @@ def Dict(keys, values):
     ret_type = TDict(key_col._type, value_col._type)
     return _func("Dict", ret_type, keys, values)
 
+def dbeta(x, a, b):
```
Add the typecheck.

jigold commented on python/hail/expr/functions.py:
```diff
@@ -640,6 +658,23 @@ def parse_locus(s, reference_genome=None):
                           s._indices, s._aggregations, s._joins, s._refs)
 
+@typecheck(pl=ArrayNumericExpression)
+def pl_dosage(pl):
```
Should this allow a python list of length 3 as well?

tpoterba replied:
this should be `expr_array` and the element type should be checked in the method

jigold commented on python/hail/expr/functions.py:
```diff
@@ -640,6 +658,23 @@ def parse_locus(s, reference_genome=None):
                           s._indices, s._aggregations, s._joins, s._refs)
 
+@typecheck(pl=ArrayNumericExpression)
+def pl_dosage(pl):
```
Does these two functions need to get added to an `__init__.py` file?

tpoterba replied:
I don't think so, since functions is a single-file module

patrick-schultz replied:
Right. I added these because the old doc examples for skat used them. I ported the examples to the new docstring, which uses both functions, and the tests pass.

patrick-schultz commented on python/hail/expr/functions.py:
```diff
@@ -640,6 +666,25 @@ def parse_locus(s, reference_genome=None):
+    :class:`.Float64Expression`
+    """
+    if not isinstance(pl, ArrayNumericExpression):
+        raise TypeError("Function 'pl_dosage' expects an 'ArrayNumericExpression'. Found {}.".format(type(pl)))
```
@tpoterba How is this? I doesn't seem to fit in with other type errors I've seen, but I couldn't think of any other way to do it.

tpoterba replied:
I think using the Expression class names is going to be very confusing for the user. How about:
```python
element_type = pl.dtype.element_type
if not is_numeric(element_type):
    raise TypeError("function 'pl_disage' expects an expression of type 'Array[numeric]', found '{}'".format(pl.dtype))
```

tpoterba replied:
not quite sure about the 'Array[numeric]' but the rest seems good

jigold commented on python/hail/expr/functions.py:
```diff
@@ -382,6 +382,32 @@ def Dict(keys, values):
+    """
+    Returns the probability density at x of a `beta distribution <https://en.wikipedia.org/wiki/Beta_distribution>`__
+    with parameters a (alpha) and b (beta).
+
```
Sorry I didn't notice this earlier. All of the other expr functions have examples. Can you put those in for both `dbeta` and `plDosage` ?

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -204,6 +204,179 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+    weights in Hail using alternate allele frequencies stored in a row-indexd
+    field `AF`, one can use the expression:
+
+    ``functions.dbeta(AF[0].min(AF[1]), 1.0, 25.0) ** 2``
```
I think this would be better as a code block `>>>` (looks nicer and will be tested to guard against future interface changes). Feel free to push back if it's too difficult to make the example work or it's not as clear with the extra text referring to `ds`.

jigold commented:
The test is broken.

patrick-schultz commented:
@jigold I fixed the tests (things got broken rebasing onto breaking, unsurprisingly). I also incorporated a change to `Table.view_join_rows` allowing for implicit joins with `Locus` and `Interval` keyed tables into `Variant` keyed datasets (cf. @cseed @tpoterba), which I needed for the skat test.

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+
+    .. testsetup:: *
+
+        ds2 = methods.variant_qc(dataset)
```
Can you add `ds2 = ds2.annotate(AF = ds2.variant_qc.AF)` to the doctest setup. Then you can simplify the expression below since `AF` will be top level.

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+    as 1 and 0, respectively.
+
+    The resulting :class:`.Table` provides the group's key, the size (number of
+    variants) in the group, the variance component score ``qstat``, the SKAT
```
rows instead of variants

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+           iterations=integral)
+def skat(dataset, key_expr, weight_expr, y, x, covariates=[], logistic=False,
+         max_size=46340, accuracy=1e-6, iterations=10000):
+    r"""Test each keyed group of variants for association by linear or logistic
```
rows instead of variants.

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+
+    .. caution::
+
+       To process a group with :math:`m` variants, several copies of an
```
rows instead of variants

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+
+       To process a group with :math:`m` variants, several copies of an
+       :math:`m \times m` matrix of doubles must fit in worker memory. Groups
+       with tens of thousands of variants may exhaust worker memory causing the
```
rows instead of variants

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+    <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3135811/>`__.
+
+    The test is run on complete samples (i.e., phenotype and all covariates
+    non-missing). For each variant, missing input values are imputed as the
```
row instead of variant

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+    `Rare-Variant Association Testing for Sequencing Data with the Sequence Kernel Association Test
+    <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3135811/>`__.
+
+    The test is run on complete samples (i.e., phenotype and all covariates
```
The test is run on columns with non-missing values for `x` and `covariates`.

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+    non-missing). For each variant, missing input values are imputed as the
+    mean of all non-missing input values.
+
+    Variant weights must be non-negative. Variants with missing weights are
```
Variants -> Rows.

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+        >>> functions.dbeta(ds2.variant_qc.AF.min(1 - ds2.variant_qc.AF),
+        ...                 1.0, 25.0) ** 2
+
+    In the logistic case, the phenotype must either be numeric (with all
```
the phenotype `x`

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+    +------+------+-------+-------+-------+
+
+    Groups larger than `max_size` appear with missing ``qstat``, ``pval``, and
+    ``fault``. The hard limit on the number of variants in a group is 46340.
```
variants -> rows

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+    Parameters
+    ----------
+    key_expr : :class:`.Expression`
+        Row-indexed expression for key associated to each variant.
```
variant -> row

jigold commented on python/hail/methods/statgen.py:
```diff
@@ -205,6 +205,186 @@ def linreg(dataset, ys, x, covariates=[], root='linreg', block_size=16):
+    key_expr : :class:`.Expression`
+        Row-indexed expression for key associated to each variant.
+    weight_expr : :class:`.NumericExpression`
+        Row-indexed expression of numeric type for variant weights.
```
row weights.

patrick-schultz commented on python/hail/methods/statgen.py:
```diff
@@ -1005,46 +1005,47 @@ def skat(dataset, key_expr, weight_expr, y, x, covariates=[], logistic=False,
-    weights in Hail using alternate allele frequencies stored in a row-indexd
-    field `AF`, one can use the expression:
+    Row weights must be non-negative. Rows with missing weights are ignored. In
+    the R package ``skat``---which assumes rows are variants---default weights
```
Is it the rendered html or the source you're objecting to? Traditionally there are no spaces around an em-dash, though I think many newspapers use that style (things I learned about typesetting working with LaTeX). But I agree that the em-dashes in the font we're using in the docs is a bit smaller than I would like. I'm okay with either style. But if it's the source you're concerned about, I think there are other ways to encode an em-dash.
